### ISSUE TITLE
TSDFVoxelGrid pointcloud generation from pinhole camera parameters (camera frustum)

### ISSUE THREAD

**Issue by VladMVLX** [intel-isl/Open3D#3491]
Hello Open3D developers,

First of all I want to say thank you for doing amazing job on library and CUDA integration.

At the moment we are doing some research on real-time scanning with cloud computing where we try to separate scanning and integration / odometry routines and run them on discrete networked machines.

Till now we are using @theNded old branch of Open3D with CUDA implementation https://github.com/theNded/Open3D ,
and trying to switch to the main branch of Open3D, as we understand old branch is not supported anymore and main branch of Open3D library has awesome tensor pipeline now with CUDA support. 

In my attempts I found that current version of Open3D in master is missing feature I used to optimize our routine: getting subvolumes that are in camera frustum. 

Using old branch I did it by calling:

`_tsdfVolume->ResetActiveSubvolumeIndices();`
`_tsdfVolume->active_subvolume_entry_array_.Clear();`
`_tsdfVolume->GetSubvolumesInFrustum(intr, extr);`
 

After I did marching cubing on active subvolumes thus I can have point cloud that can be used later for ICP registration of next frame and point cloud synchronization between network nodes.

It seams there are no any functionality in current master branch of Open3D that can do so.

Can you please advise me if I am missing something?
Is there any way I can emulate old behavior or get TSDF part that are in frustum and generate point cloud using it?

**Comment by theNded** [comment]
Hi, now I'm experimenting with a new branch: https://github.com/intel-isl/Open3D/tree/wei/tsdf-experiment.

In this branch I'm using a hashmap for points in the frustum (`point_hashmap`). It produces a vertex map that can be viewed as a point cloud in the frustum: https://github.com/intel-isl/Open3D/blob/wei/tsdf-experiment/cpp/open3d/t/geometry/TSDFVoxelGrid.cpp#L284

You can refer to this line to obtain a vertex map: https://github.com/intel-isl/Open3D/blob/wei/tsdf-experiment/examples/cpp/TIntegrateRGBD.cpp#L196

and these lines to convert it to a point cloud: https://github.com/intel-isl/Open3D/blob/wei/tsdf-experiment/cpp/open3d/t/geometry/PointCloud.cpp#L298-L326

**Comment by VladMVLX** [comment]
Hello @theNded,
Thank you for the directions, will try to follow your advice

**Comment by theNded** [comment]
Closed for now. Please reopen if there are further questions. I will also keep it on my task list.

**Comment by VladMVLX** [comment]
Hello @theNded ,

I tried approach you recommended on master branch (as it seams your branch was merged into the master in v0.13)

I am getting extrinsics matrix from T265 tracking camera for integrate and raycasting function calls:

```c++
open3d::core::Tensor extrinsics = open3d::core::eigen_converter::EigenMatrixToTensor(Transform);
auto inverse_extrinsics = open3d::t::geometry::InverseTransformation(extrinsics);

_tsdfVolume->Integrate(depth, color, intrinsics, inverse_extrinsics, depthScale, maxDepth);


auto result = _tsdfVolume->RayCast(
	intrinsics, inverse_extrinsics, width, height,
	1, 0.0f, maxDepth, 3.0f,
	TSDFVolume::SurfaceMaskCode::ColorMap | TSDFVolume::SurfaceMaskCode::VertexMap | TSDFVolume::SurfaceMaskCode::DepthMap);
```

getting VertexMap:

`open3d::t::geometry::Image vertexImg(result[TSDFVolume::SurfaceMaskCode::VertexMap]);`

and converting it into point cloud as you recommended:

```c++
const int64_t im_size = vertexImg.GetRows() * vertexImg.GetCols();

const float invalid_fill = NAN;

auto vertex_list = vertexImg.AsTensor().View({ im_size, 3 });

f32_extrinsics= f32_extrinsics.Inverse();

vertex_list = vertex_list
		.Matmul(f32_extrinsics.Slice(0, 0, 3, 1)
		.Slice(1, 0, 3, 1))
		.Add_(f32_extrinsics.Slice(0, 0, 3, 1)
		.Slice(1, 3, 4, 1)
		.T());

vertexImg = open3d::t::geometry::Image(vertex_list.View(vertexImg.AsTensor().GetShape())
			.Contiguous());
		
auto normal_map_t = vertexImg.CreateNormalMap(invalid_fill)
		.AsTensor()
		.View({ im_size, 3 });
		// all columns are the same
auto valid_idx =
	normal_map_t.Slice(1, 0, 1, 1)
	.IsFinite()
	.LogicalAnd(vertex_list.Slice(1, 0, 1, 1).IsFinite())
	.Reshape({ im_size });
open3d::t::geometry::PointCloud pcd(
			{ {"points",
			  vertex_list.GetItem({open3d::core::TensorKey::IndexTensor(valid_idx),
								   open3d::core::TensorKey::Slice(open3d::core::None, open3d::core::None, open3d::core::None)})},
			 {"normals",
			  normal_map_t.GetItem({open3d::core::TensorKey::IndexTensor(valid_idx),
									open3d::core::TensorKey::Slice(open3d::core::None, open3d::core::None, open3d::core::None)})} });
if (colorImg.GetRows() > 0) {
	pcd.SetPointColors(
		colorImg.AsTensor()
		.View({ im_size, 3 })
		.GetItem({ open3d::core::TensorKey::IndexTensor(valid_idx),
				  open3d::core::TensorKey::Slice(open3d::core::None, open3d::core::None, open3d::core::None) }));
		}
```
Everything works fine, and I am getting pretty accurate point cloud from camera frustum view while camera standing stationary with only rotations applied. 

The problem appears when camera moved physically along any of direction (X,Y,Z) generated point cloud from vertex map seams to have wrong displacement and not aligning with point cloud from pervious frames:
![image](https://user-images.githubusercontent.com/58504993/122222880-5bd7af00-cebb-11eb-9257-11fd5b65156a.png)

I also tried to use:

`open3d::t::geometry::PointCloud::CreateFromRGBDImage(rgbd_image, intrinsics, f32_extrinsics, 1.f, maxDepth, 1, true);`

But results seams to be the same.

This problem only appears on point clouds generated from vertex maps in real-time, the whole point cloud extracted by:
```c++
tsdfVolume->ExtractSurfacePoints(
			3000000, 0.0f,
			TSDFVolume::SurfaceMaskCode::VertexMap | TSDFVolume::SurfaceMaskCode::ColorMap | TSDFVolume::SurfaceMaskCode::NormalMap);
```
Looks fine and don`t have any visual align related artifacts:
![image](https://user-images.githubusercontent.com/58504993/122222909-61cd9000-cebb-11eb-9439-0a52314318fb.png)

I suspect there are some minor issue in vertex map generation or code that converts it to the point cloud which resulting in wrong point cloud displacement, can you please advise?

Sorry for bad quality scans, made them only for debug purposes, but hopefully they are well representing the problem.
